### ISSUE TITLE
Concurrent read segfault

### ISSUE THREAD

**Issue by ArnaudLevaufre** [Unidata/netcdf4-python#844]
Hello

When openning a netcdf file in read mode more than one time in a ThreadPool results in a segfault. When using a process pool there is no issue and the concurrent read works fine. It can be an issue in a web environment where the server is using threads to handle its clients and two or more clients make a request that need to read the same netcdf file.

I have a simple python script and a dataset that reproduces the segfault:
[netcdfSegfault.zip](https://github.com/Unidata/netcdf4-python/files/2345151/netcdfSegfault.zip)

The python script provided in the zip:
```python
import netCDF4
from multiprocessing.pool import ThreadPool

def read_netcdf(path):
    ieast = 1750
    iwest = 1760
    inorth = 1380
    isouth = 1370

    with netCDF4.Dataset(path, 'r') as ncf:
        return ncf['/Depth/ndepths'][:][iwest:ieast, isouth:inorth]


if __name__ == "__main__":
    print("Netcdf4 version: %s" % netCDF4.__version__)
    path = "./max_depth_ndepth_quonops.nc"
    with ThreadPool(2) as p:
        print(p.map(read_netcdf, [path for i in range(2)]))

```

Will output the following when executed.
```
Netcdf4 version: 1.4.0
Segmentation fault (core dumped)
```

**Comment by jswhit** [comment]
May be related to https://github.com/Unidata/netcdf4-python/issues/640

**Comment by jswhit** [comment]
Works with python3.6 if you replace `ncf['/Depth/ndepths'][:][iwest:ieast, isouth:inorth]` with `ncf['/Depth/ndepths'][iwest:ieast, isouth:inorth]`.  I think it's a memory issue, not a concurrency issue.

**Comment by shoyer** [comment]
You need to use a lock when using netCDF4 with multiple threads. Unfortunately, the underlying HDF5 library is not thread safe.

**Comment by dopplershift** [comment]
Yeah, my rule from experience is that if I'm using multiple threads, all access to netcdf4-python needs to be guarded by a lock.

**Comment by jswhit** [comment]
You can build HDF5 thread-safe, but it's not the default
https://portal.hdfgroup.org/display/knowledge/Questions+about+thread-safety+and+concurrent+access

**Comment by dopplershift** [comment]
Even if HDF5 is compiled with thread safety, the netcdf4 C library is not thread safe.

**Comment by ArnaudLevaufre** [comment]
Thanks for the answers. I didn't know the HDF5 library was not thread safe (I must admit I didn't look for this information and took it for granted). @jswhit removing the middle `[:]` works on the example but not reliably with the real implementation in the context of a web server. So for now I will stick to using processes for concurent read.

**Comment by jswhit** [comment]
There was at least a plan to make the netcdf library thread safe (see https://www.unidata.ucar.edu/blogs/developer/entry/implementing-thread-safe-access-to).  @DennisHeimbigner - was that every implemented?

**Comment by DennisHeimbigner** [comment]
Not yet. It is waiting on some other library changes. The priority is high, however.

**Comment by lanougue** [comment]
Hi everyone
@ArnaudLevaufre , did you have experienced similar errors with concurent reads even with processes ?
I opened an issue in xarray ([issue](https://github.com/pydata/xarray/issues/2494)) since I have strong issues trying to read netcdf files with several processes.
Looking for some help here too ...
Thks
